### PR TITLE
LFX Phase 1: Ground truth map for architectural parameter extraction

### LINKED ISSUE #1747: LFX - Phase 1: Build the Ground Truth Map from UDB Parameters
## Objective

Create a machine-readable catalog of all 185 existing UDB architectural parameters, enriched with schema analysis, CSR cross-references, heuristic classifications, and candidate spec text locations. This catalog serves as both the training data for LLM prompts and the validation benchmark for evaluating LLM extraction accuracy.

## Deliverables

| Deliverable | Description |
|---|---|
| `ground_truth.json` | Full structured data for all 185 parameters: name, description, value type analysis, definedBy conditions, CSR cross-references, classification with confidence and reasoning |
| `spec_mappings.json` | For each parameter, the top candidate locations in the spec `.adoc` files with relevance scores, line numbers, and text context |
| `parameters_catalog.csv` | Spreadsheet-ready catalog (19 columns) for mentor review |
| `phase1_report.txt` | Human-readable report with statistics and per-parameter breakdown |
| `udb_param_names.txt` | Flat list of 185 parameter names for inclusion in LLM prompts |

### Step 1.1 — Export UDB parameters to structured JSON
- Read all `spec/std/isa/param/*.yaml` files (excluding MOCK fixtures)
- Derive value type programmatically from JSON Schema (boolean -> binary, integer+enum -> enum, integer+min/max -> range, array -> set/bitmask, oneOf+when -> conditional)
- Parse complex `definedBy` structures (allOf, anyOf, parameter conditions)
- Cross-reference all CSR YAML files to find IDL code (`sw_write()`, `type()`, `reset_value()`, `legal?()`) that references each parameter

### Step 1.2 — Map parameters to spec text locations
- Search all 74 `.adoc` files (52,602 lines) using multi-strategy keyword matching
- Search strategies: exact parameter name, CSR name in backticks, description keywords, WARL+CSR proximity patterns, name segment matching
- Score and rank candidate lines; extract surrounding context
- Exclude non-normative content (NOTE/TIP/WARNING blocks)

### Step 1.3 — Classify every UDB parameter
Heuristic classification into:
- **NORM_DIRECT** — Directly configurable, not CSR-controlled (e.g., `MXLEN`, `NUM_PMP_ENTRIES`)
- **NORM_CSR_WARL** — Legal values of a WARL CSR field (e.g., `MTVEC_MODES`, `SATP_MODE_BARE`)
- **NORM_CSR_RW** — Controls whether a CSR field is RO/RW (e.g., `MUTABLE_MISA_C`, `M_MODE_ENDIANNESS`)
- **SW_RULE** — Software-deterministic with correct fencing (e.g., `HW_MSTATUS_FS_DIRTY_UPDATE`)

## Acceptance Criteria
- [ ] All 185 non-MOCK parameters exported with complete metadata
- [ ] Value types verified against actual YAML schemas (100% match)
- [ ] CSR cross-references verified against actual CSR YAML files
- [ ] >= 95% of parameters have at least one spec text candidate match
- [ ] Classification confidence breakdown: >= 75% high confidence
- [ ] CSV catalog ready for mentor review

### PR DESCRIPTION
**Pull request by ishaan-arora-1**
## Summary

- Adds scripts and data that catalog all 185 UDB architectural parameters with schema analysis, CSR cross-references, heuristic classifications, and candidate spec text locations
- This is the foundation for LLM-based parameter extraction from the RISC-V privileged and unprivileged specifications (see #1747)
- Part of the LFX project to systematically identify and tag architectural parameters in the spec

### Scripts (`param_extraction/scripts/`)

| Script | Purpose |
|---|---|
| `export_udb_params.py` | Reads all 185 `spec/std/isa/param/*.yaml` files (excluding 22 MOCK fixtures), analyzes JSON Schema structure, cross-references CSR IDL code for `sw_write()`/`type()`/`reset_value()` references, and classifies each parameter |
| `map_params_to_spec.py` | Searches all 74 spec `.adoc` files (52,602 lines) for text related to each parameter using multi-strategy keyword matching (exact name, CSR backtick refs, description keywords, WARL proximity patterns) |
| `generate_report.py` | Produces the CSV catalog, text report, and flat parameter name list |

### Data outputs (`param_extraction/data/`)

| File | Description |
|---|---|
| `ground_truth.json` | Full structured data for all 185 parameters: name, description, value type, definedBy, CSR cross-references, classification with confidence and reasoning |
| `spec_mappings.json` | Top candidate spec text locations per parameter with relevance scores, line numbers, and context |
| `parameters_catalog.csv` | 19-column spreadsheet-ready catalog |
| `phase1_report.txt` | Human-readable report with statistics and per-parameter breakdown |
| `udb_param_names.txt` | Flat list of 185 parameter names (for inclusion in LLM prompts in later phases) |

## Key results

| Metric | Value |
|---|---|
| Parameters cataloged | 185 (22 MOCK fixtures excluded) |
| Classification: NORM_DIRECT | 102 (55%) — directly configurable, not CSR-controlled |
| Classification: NORM_CSR_RW | 55 (30%) — controls RO/RW behavior of CSR fields |
| Classification: NORM_CSR_WARL | 26 (14%) — legal values of WARL CSR fields |
| Classification: SW_RULE | 2 (1%) — software-deterministic with correct fencing |
| High-confidence classifications | 150 (81%) |
| Value type: binary | 111 (60%), enum: 36 (19%), range: 12 (6%) |
| Parameters with CSR cross-references | 94 (51%) |
| Parameters mapped to spec text | 183/185 (98%) |
| Strong spec matches (score >= 5) | 161 (87%) |

## How to run

```bash
# Requires PyYAML (pip install pyyaml)
# Requires ext/riscv-isa-manual submodule to be initialized

python3 param_extraction/scripts/export_udb_params.py
python3 param_extraction/scripts/map_params_to_spec.py
python3 param_extraction/scripts/generate_report.py
```

## Test plan

- [x] All 185 non-MOCK parameters exported with complete metadata
- [x] Value types verified against actual YAML schemas (100% match)
- [x] CSR cross-references verified against actual CSR YAML files
- [x] 98% of parameters have at least one spec text candidate match
- [x] 81% high-confidence classifications (target was >= 75%)
- [x] CSV catalog and JSON outputs are consistent (all 185 rows match)
- [x] No duplicate parameter names
- [x] All source YAML files exist on disk
- [x] Scripts run cleanly end-to-end with no errors

Closes #1747

---

## Update — review tweaks applied

Went back through the review and pushed the two small `generate_report.py` cleanups:

- Added `encoding="utf-8"` to the `open(names_path, "w")` call so the names file is written with an explicit text encoding instead of relying on the platform default.
- Added an `if not rows: return` guard before the CSV write so we don't blow up on `rows[0].keys()` if there's nothing to write.

No data regen for this phase — the catalog files are byte-identical for any non-empty dataset (UTF-8 is already the default on Linux/macOS and `rows` is never empty for real input).

### CODE REVIEW

File: param_extraction/scripts/generate_report.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+                "spec_in_note": best.get("in_note", ""),
+                "spec_text": best.get("line_text", "")[:200],
+                "description": p["description"][:300],
+            }
+        )
+
+    with open(out_path, "w", newline="", encoding="utf-8") as f:
+        writer = csv.DictWriter(f, fieldnames=list(rows[0].keys()))
+        writer.writeheader()
+        writer.writerows(rows)
```
Comment (miguelcsx) [inline param_extraction/scripts/generate_report.py:91]:
```suggestion
    if not rows:
        return

    with open(out_path, "w", newline="", encoding="utf-8") as f:
        writer = csv.DictWriter(f, fieldnames=list(rows[0].keys()))
        writer.writeheader()
        writer.writerows(rows)
```

File: param_extraction/scripts/generate_report.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    with open(DATA_DIR / "spec_mappings.json", encoding="utf-8") as f:
+        sm = json.load(f)
+
+    params = gt["parameters"]
+    mappings = {m["parameter_name"]: m for m in sm["mappings"]}
+
+    # Generate flat name list (for LLM prompts)
+    names = sorted(p["name"] for p in params)
+    names_path = DATA_DIR / "udb_param_names.txt"
+    with open(names_path, "w") as f:
```
Comment (miguelcsx) [inline param_extraction/scripts/generate_report.py]:
```suggestion
    with open(names_path, "w", encoding="utf-8") as f:
```
